base_layer/core/src/base_node/state_machine_service/states/horizon_state_sync/error.rs

```diff
@@ -22,44 +22,45 @@
 
 use crate::{
     base_node::{comms_interface::CommsInterfaceError, state_machine_service::states::helpers::BaseNodeRequestError},
-    chain_storage::ChainStorageError,
+    chain_storage::{ChainStorageError, MmrTree},
     transactions::transaction::TransactionError,
+    validation::ValidationError,
 };
+use tari_comms::protocol::rpc::{RpcError, RpcStatus};
+use tari_mmr::error::MerkleMountainRangeError;
 use thiserror::Error;
 use tokio::task;
 
 #[derive(Debug, Error)]
 pub enum HorizonSyncError {
     #[error("Peer sent an empty response")]
     EmptyResponse,
-    #[error("Peer sent an invalid response")]
-    IncorrectResponse,
+    #[error("Peer sent an invalid response: {0}")]
+    IncorrectResponse(String),
     #[error("Exceeded maximum sync attempts")]
     MaxSyncAttemptsReached,
     #[error("Chain storage error: {0}")]
     ChainStorageError(#[from] ChainStorageError),
     #[error("Comms interface error: {0}")]
     CommsInterfaceError(#[from] CommsInterfaceError),
-    // #[error("Final state validation failed: {0}")]
-    // FinalStateValidationFailed(ValidationError),
+    #[error("Final state validation failed: {0}")]
+    FinalStateValidationFailed(ValidationError),
     #[error("Join error: {0}")]
     JoinError(#[from] task::JoinError),
     #[error("Invalid kernel signature: {0}")]
     InvalidKernelSignature(TransactionError),
-    // #[error("Validation failed for {0} MMR")]
-    // InvalidMmrRoot(MmrTree),
+    #[error("Validation failed for {0} MMR")]
+    InvalidMmrRoot(MmrTree),
+    #[error("Invalid range proof for output:{0} : {1}")]
+    InvalidRangeProof(String, String),
     #[error("Base node request error: {0}")]
     BaseNodeRequestError(#[from] BaseNodeRequestError),
-}
-
-impl HorizonSyncError {
-    pub fn is_recoverable(&self) -> bool {
-        // use HorizonSyncError::*;
-        unimplemented!()
-        // match self {
-        // FinalStateValidationFailed(_) | InvalidMmrRoot(_) => false,
-        //  InvalidMmrRoot(_) => false,
-        // _ => true,
-        // }
-    }
+    #[error("RPC error: {0}")]
+    RpcError(#[from] RpcError),
+    #[error("RPC status: {0}")]
+    RpcStatus(#[from] RpcStatus),
```

Comment:
I am a bit confused by this error looking it at. 
Why do we have two error types here? just looking at the names I would be let to believe that RpcError is a possible state of RpcStatus?

Or are we talking completely different parts of the system here?

Reply:
`RpcError` is an error type for the comms RPC module. `RpcStatus` encapsulates the status of responses sent back to clients (OK, Bad Request, General error etc), it contains a status code and a message. 

Perhaps `RpcRequestFailed(#[from] RpcStatus)` would be a more descriptive name